src/fortran/fdb.f90

```diff
@@ -1,39 +1,329 @@
+      integer(kind=c_int) function fdb_listiterator_attrs(it, uri, off, len) bind(C,name='fdb_listiterator_attrs')
+         use, intrinsic :: iso_c_binding, only : c_int, c_ptr, c_int, c_char
+         type(c_ptr), intent(in), value      :: it
+         character(kind=c_char, len=1), dimension(*), INTENT(inout) :: uri
+         integer(kind=c_int), intent(inout) :: off
+         integer(kind=c_int), intent(inout) :: len
+      end function fdb_listiterator_attrs
+   end interface
+
+   CONTAINS
+   SUBROUTINE get_value_of_key(igrib, keyname, keyvalue_str, keyvalue_int)
+      character(len=*), INTENT(IN)         :: keyname
+      character(len=128), INTENT(OUT)    :: keyvalue_str
+      integer, INTENT(INOUT), OPTIONAL     :: keyvalue_int
+      integer, INTENT(IN)                 :: igrib
+      integer                             :: res
+      integer                             :: is_missing
+      call codes_is_missing(igrib, trim(keyname), is_missing);
+      if (is_missing /= 1) then
+         ! key value is not missing so get values
+         if (present(keyvalue_int)) then
+            call codes_get_int(igrib, keyname, keyvalue_int)
+            write (keyvalue_str, "(I4)") keyvalue_int
+            keyvalue_str=trim(adjustl(keyvalue_str))
+         else
+            call codes_get(igrib, keyname, keyvalue_str)
+         end if
+         write (*, *) keyname, '=', keyvalue_str
+      else
+         write (*, *) keyname, ' is missing from data.'
+      end if
+   END SUBROUTINE
+
+   SUBROUTINE copy_s2a(a,s)   ! copy s(1:Clen(s)) to char array
+      CHARACTER(*),INTENT(IN) :: s
+      CHARACTER :: a(LEN(s)+1)
+      INTEGER :: i
+      DO i = 1,LEN(s)
+         a(i) = s(i:i)
+      END DO
+      a(LEN(s)+1) = char(0)
+   END SUBROUTINE copy_s2a
+
+   SUBROUTINE copy_a2s(a,s)  
+      CHARACTER,INTENT(IN) :: a(:)
+      CHARACTER(len=*),INTENT(INOUT)  :: s
+      INTEGER :: i
+      DO i = 1,SIZE(a)
+         s(i:i) = 'z' ! a(i) ! FAILS HERE
+      END DO
+   END SUBROUTINE copy_a2s
+
+   FUNCTION index_chararray(a,s) result(idx)  
+      CHARACTER,INTENT(IN) :: a(:)
+      CHARACTER(len=*), INTENT(IN)  :: s
+      CHARACTER :: s_a(LEN(s))
+      INTEGER  :: i, j, idx
+      logical :: equal
+      DO i = 1,LEN(s)
+         s_a(i) = s(i:i)
+      END DO
+      DO i = 1,SIZE(a)-SIZE(s_a)
+         DO j = 1, SIZE(s_a)
+            equal = a(i+j-1) == s_a(j)
+            IF (.not. equal) THEN
+               exit
+            END IF
+         END DO
+         IF (equal) THEN
+            idx=i
+            exit
+         END if
+      END DO
+      IF (.not. equal) THEN
+         idx=0
+      END IF
+   END FUNCTION index_chararray
+   
+   SUBROUTINE add_key(key, igrib, keyname_str, type)  
+      use, intrinsic :: iso_c_binding, only : c_int, c_ptr, c_char
+      type(c_ptr), INTENT(INOUT)                  :: key
+      character(len=*), INTENT(IN)                :: keyname_str
+      character(len=*), INTENT(IN), OPTIONAL      :: type
```

Comment:
you could also use enum ? although is Fortran 2003, so not sure, just something to consider since it will be faster than string comparisons